base/reduce.jl

```diff
@@ -302,8 +269,40 @@ implementations may reuse the return value of `f` for elements that appear multi
 `itr`. Use [`mapfoldl`](@ref) or [`mapfoldr`](@ref) instead for
 guaranteed left or right associativity and invocation of `f` for every value.
 """
-mapreduce(f, op, itr; kw...) = mapfoldl(f, op, itr; kw...)
-mapreduce(f, op, itr, itrs...; kw...) = reduce(op, Generator(f, itr, itrs...); kw...)
+mapreduce(f::F, op::G, x; init=_InitialValue()) where {F,G} = mapreduce_pairwise(f, op, x, init)
+mapreduce(f::F, op::G, x, xs...; init=_InitialValue()) where {F,G} = mapreduce_pairwise(identity, op, Generator(f, x, xs...), init)
+
+"""
+    mapreduce(f, op, A::AbstractArray...; dims=:, [init])
+
+Evaluates to the same as `reduce(op, map(f, A...); dims=dims, init=init)`, but is generally
+faster because the intermediate array is avoided.
+
+!!! compat "Julia 1.2"
+    `mapreduce` with multiple iterators requires Julia 1.2 or later.
+
+# Examples
+```jldoctest
+julia> a = reshape(Vector(1:16), (4,4))
+4×4 Matrix{Int64}:
+ 1  5   9  13
+ 2  6  10  14
+ 3  7  11  15
+ 4  8  12  16
+
+julia> mapreduce(isodd, *, a, dims=1)
+1×4 Matrix{Bool}:
+ 0  0  0  0
+
+julia> mapreduce(isodd, |, a, dims=1)
+1×4 Matrix{Bool}:
+ 1  1  1  1
+```
+"""
+mapreduce(f::F, op::G, A::AbstractArrayOrBroadcasted; init=_InitialValue(), dims=(:)) where {F,G} = mapreducedim(f, op, A, init, dims)
+mapreduce(f::F, op::G, A::AbstractArrayOrBroadcasted, As::AbstractArrayOrBroadcasted...; init=_InitialValue(), dims=(:)) where {F,G} =
+    reduce(op, map(f, A, As...); init, dims)
+mapreducedim(f::F, op::G, A, init, ::Colon) where {F,G} = mapreduce_pairwise(f, op, A, init)
```

Comment:
I wonder (suspect?) if smaller `pairwise_blocksize` is better when the cost of `f` exceeds the cost of `op`? e.g. to start with:
```
pairwise_blocksize(::typeof(sqrt), op) = 64
pairwise_blocksize(::typeof(sin), op) = 64
pairwise_blocksize(::typeof(cos), op) = 64
pairwise_blocksize(::typeof(exp), op) = 64
pairwise_blocksize(::typeof(log), op) = 64
```